### PR TITLE
EIP-1967: Standard Proxy Storage Slots

### PR DESCRIPTION
**Pull request by spalladino**
We are proposing an EIP to standardise how proxies store the address of the logic contract they delegate to. Given that the delegating proxy contract pattern has become widespread, we believe there is value, especially for off-chain tooling and explorers, in having a standard storage layout for proxy-specific information. In particular, a storage layout based on the [unstructured storage pattern](https://blog.zeppelinos.org/upgradeability-using-unstructured-storage/).

We are pushing for a standard storage layout and not a standard interface since proxies are designed to act transparently for a user, and introducing proxy-specific functions can [lead to attacks](https://medium.com/nomic-labs-blog/malicious-backdoors-in-ethereum-proxies-62629adf3357).

This EIP is designed to be a generalization for other delegating proxy standards that use the unstructured storage pattern, such as [EIP-1822](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1822.md).

Discussions at [ethereum-magicians forum](https://ethereum-magicians.org/t/eip-nnnn-standard-proxy-storage-slots/3185).

### CODE REVIEW

**Comment by wjmelements** [comment]
I agree that being able to find implementation addresses is important, particularly for discovering the underlying contract ABI.

TrustToken's proxies do not store their implementation or owner at those OpenZeppelin slots. For them, the proper way to read the proxy is by calling `implementation` and `proxyOwner`. A public interface is our preferred solution, and should be tried alongside `getStorageAt` by services seeking implementation addresses.

Here are the current TrustToken proxies:
* 0x0000000000013949F288172bD7E36837bDdC7211
* 0x0000000000085d4780B73119b644AE5ecd22b376
* 0x0000000000075EfBeE23fe2de1bd0b7690883cc9
* 0x00000000441378008EA67F4284A57932B1c000a5
* 0x00000000BbcF7700A1b403C9EB666f350707b900
* 0x00006100f7090010005f1bd7ae6122c3c2cf0090
* 0x0000109a8344DE9c00465264006C0000769A2770
* 0x00000100F2A2bd000715001920eB70D229700085
* 0x00005cAD001e0900002979f7314D00Fc480a29bD

[Nomic](https://medium.com/nomic-labs-blog/malicious-backdoors-in-ethereum-proxies-62629adf3357)'s suggested solution for proxy method ID collision is unsatisfactory; the proxy owner can still be fooled into invoking the malicious collision. Their fix only makes it harder to detect in testing, or rather, easier to hide. It also creates issues for several third-party contract readers, such as Etherscan, since they cannot access the privileged read methods.

Aside, I find that cementing proxy administrative methods in the proxy prevents categories of permanent operational failures. For example, if the function that upgrades the proxy is only present in the implementation contract, it is possible for the proxy owner to upgrade to the wrong address (e.g. 0x0) or a buggy implementation but never been able to correct it. It is difficult to ascertain on-chain that a new implementation can manage the proxy correctly.

Lastly, I challenge the idea that `0x7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c3` is the best storage location. It has a known keccak preimage, unlike nearby `0x7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c2`. It is harder to engineer an attack without a known preimage.

**Comment by spalladino** [comment]
Thanks for the comments @wjmelements!

> [Nomic](https://medium.com/nomic-labs-blog/malicious-backdoors-in-ethereum-proxies-62629adf3357)'s suggested solution for proxy method ID collision is unsatisfactory; the proxy owner can still be fooled into invoking the malicious collision.

Note that the proxy owner can _only_ call proxy administrative methods, which makes it more difficult to fool them, since the proxy does not delegate the call to the logic if it came from the admin.

> For example, if the function that upgrades the proxy is only present in the implementation contract, it is possible for the proxy owner to upgrade to the wrong address (e.g. 0x0) or a buggy implementation but never been able to correct it. It is difficult to ascertain on-chain that a new implementation can manage the proxy correctly.

Agree there, that's why the details of where the upgrade function is placed are not part of this EIP, but of EIP 1822.

> Lastly, I challenge the idea that `0x7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c3` is the best storage location. It has a known keccak preimage, unlike nearby `0x7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c2`. It is harder to engineer an attack without a known preimage.

Good point. We chose a hash with such a preimage so that we knew that it would not collide with anything allocated by the Solidity or Vyper compilers (they both store mapping values in a preimage that starts with the slot number where the mapping is declared), but picking an entirely random value could be a good idea as well. Curious to see what others think about this!

File: EIPS/eip-1967.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+Storage slot `0x7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c3`
+(obtained as `keccak256('org.zeppelinos.proxy.implementation')`).
+
+Holds the address of the logic contract that this proxy delegates to.
+
+### Admin address
+
+Storage slot `0x10d6a54a4754c8869d6886b5f5d7fbfa5b4522237ea5c60d11bc4e7a1ff9390b`
+(obtained as `keccak256('org.zeppelinos.proxy.admin')`).
```
Comment (axic) [inline EIPS/eip-1967.md]: If this is a standard outside of ZeppelinOS as well, should use paths with zeppelinos?

File: EIPS/eip-1967.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+As such, the proposed storage slots for proxy-specific information are the following. They are chosen after the ones used in ZeppelinOS, which have been already deployed to mainnet by several projects, and are guaranteed to not clash with state variables allocated by the compiler. It is open for discussion to choose different slots, but we have suggested these based on current usage.
+
+More slots for additional information can be added in subsequent EIPs as needed.
```
Comment (axic) [inline EIPS/eip-1967.md]: ERCs?

File: EIPS/eip-1967.md (hunk 1 above)
Comment (spalladino) [reply EIPS/eip-1967.md]: The main reason for pushing a path with zOS was to provide support for all the proxies that have been already deployed; but if this is a deal-breaker for getting the EIP to pass, I'm fine with changing it.

Comment (axic) [reply EIPS/eip-1967.md]: I think if it uses zOS paths the title should reflect that, such as "Standard ZeppelinOS Proxy Storage Slots`.

**Comment by axic** [comment]
@wjmelements can you please comment your concerns on the discussion-to URL as well?

**Comment by wjmelements** [comment]
Not right now; there seems to be a DNS issue
Screen Shot 2019-05-16 at 5 44 14 PM

**Comment by axic** [comment]
@wjmelements it should be up now :)

@spalladino do you expect to apply any of the changes discussed above or should it be just merged as is ?

**Comment by spalladino** [comment]
> @spalladino do you expect to apply any of the changes discussed above or should it be just merged as is ?

I noted that I had a PR pending to update those on our ZeppelinSolutions fork. I've just merged it, so you should see the suggestions applied.

**Comment by axic** [comment]
@spalladino thanks I think this looks great now and satisfies EIP-1 requirements.

**Comment by spalladino** [comment]
Hey @axic! We are looking into moving the Proxy contracts into the `@openzeppelin/contracts` library, to make them easier to use without having to necessarily go through the `oz` CLI. With that in mind, it'd be good to have this EIP finalized. What do we need to do for that to happen? Thanks!

**Comment by mudgen** [comment]
@wjmelements Are there any known/successful preimage hash collision/attacks for a keccak256? 

I mean someone able to find a value that hashes to the same hash as another different value.  

Because if that is the case then aren't all Solidity mappings and dynamic arrays and strings less secure?  Because they use known preimages.

**Pull request merged**